src/utils.jl

```diff
@@ -1,63 +1,74 @@
-"""
-    local_variables = locals(frame::JuliaStackFrame)::Vector{Variable}
+## Simple utils
 
-Return the local variables as a vector of `Variable`[@ref].
-"""
-function locals(frame::JuliaStackFrame)
-    vars = Variable[]
-    for (sym, idx) in frame.last_reference
-        push!(vars, Variable(something(frame.locals[idx]), sym, false))
-    end
-    reverse!(vars)
-    if frame.code.scope isa Method
-        syms = sparam_syms(frame.code.scope)
-        for i in 1:length(syms)
-            push!(vars, Variable(frame.sparams[i], syms[i], true))
-        end
-    end
-    return vars
-end
+# Note: to avoid dynamic dispatch, many of these are coded as a single method using isa statements
```

Comment:
Won't the methods be specialized on anyway? So e.g. `scopeof` would be specialized and dynamically dispatched to so that the method `scopeof(::Module) = true` is generated after constant folding. So, is `@nospecialize` needed to prevent dynamic dispatch?

Reply:
If it can resolve it by union-splitting then dynamic dispatch won't harm anything. It's cases where it doesn't use union-splitting that we have to worry about. But you're right, it might be better to mark these anyway.

Reply:
Sure but in the union splitting case, writing the methods out in a dispatch fashion (not using `isa`) would be equivalent, no? So my point was, it is only in the case of non-specialization that the comment on this line is true?